src/atomic_densities.jl

```diff
@@ -0,0 +1,241 @@
+abstract type GuessDensityMethod end
+
+struct RandomGuessDensity <: GuessDensityMethod end
+struct GaussianGuessDensity <: GuessDensityMethod end
+struct PspGuessDensity <: GuessDensityMethod end
+struct AutoGuessDensity <: GuessDensityMethod end
+
+@doc raw"""
+    guess_density(basis::PlaneWaveBasis, method::GuessDensityMethod,
+                  magnetic_moments=[])
+
+Build a superposition of atomic densities (SAD) guess density.
+
+The guess atomic densities are taken as one of the following depending on the input
+`method`:
+
+-`RandomGuessDensity()`: A random density, normalized to the number of electrons
+`basis.model.n_electrons`. Does not support magnetic moments.
+-`AutoGuessDensity()`: A combination of the `:gaussian` and `:psp` methods where
+elements whose pseudopotentials provide numeric valence charge density data use them and
+elements without use Gaussians.
+-`GaussianGuessDensity()`: Gaussians of length specified by `atom_decay_length`
+normalized for the correct number of electrons:
+```math
+\hat{ρ}(G) = Z \exp\left(-(2π \text{length} |G|)^2\right)
+```
+- `PspGuessDensity()`: Numerical pseudo-atomic valence charge densities from the
+pseudopotentials. Will fail if one or more elements in the system has a pseudopotential
+that does not have valence charge density data.
+
+When magnetic moments are provided, construct a symmetry-broken density guess.
+The magnetic moments should be specified in units of ``μ_B``.
+"""
+function guess_density(basis::PlaneWaveBasis{T};
+                       method::GuessDensityMethod=AutoGuessDensity(),
+                       magnetic_moments=[], n_electrons=basis.model.n_electrons) where {T}
+    return _guess_density(basis, method, magnetic_moments, n_electrons)
+end
```

Comment:
Add a function `guess_density(basis, magnetic_moments) = guess_density(basis, AutoGuessDensity(), magnetic_moments)`

Re @antoine-levitt and the kwargs question. I guess one could argue to make method a kwarg, but certainly not magnetic_moments as it lives on the same footing as `basis` (data to setup the guess). The issue with making `method` a kwarg is that you cannot dispatch on it. In the common solve interface people seem to go the route of having the method as a second arg.